planner/core/push_down_predicate_to_table_scan.go

```diff
@@ -0,0 +1,305 @@
+ */
+func removeSpecificExprsFromSelection(physicalSelection *PhysicalSelection, exprs []expression.Expression) {
+	newConds := make([]expression.Expression, 0, len(physicalSelection.Conditions))
+	for _, cond := range physicalSelection.Conditions {
+		if !expression.Contains(exprs, cond) {
+			newConds = append(newConds, cond)
+		}
+	}
+	physicalSelection.Conditions = newConds
+}
+
+/**
+ * @brief: This function is used to push down the some filter conditions of the selection to the tablescan.
+ * @param: sctx: the session context
+ * @param: physicalSelection: the PhysicalSelection containing the conditions to be pushed down
+ * @param: physicalTableScan: the PhysicalTableScan to be pushed down to
+ */
+func predicatePushDownToTableScanImpl(sctx sessionctx.Context, physicalSelection *PhysicalSelection, physicalTableScan *PhysicalTableScan) {
+	// When the table is small, there is no need to push down the conditions.
+	if physicalTableScan.tblColHists.Count <= tiflashDataPackSize {
+		return
+	}
+	conds := physicalSelection.Conditions
+	if len(conds) == 0 {
+		return
+	}
+
+	// group the conditions by columns and sort them by selectivity
+	sortedConds := groupByColumnsSortBySelectivity(sctx, conds, physicalTableScan)
+
+	selectedConds := make([]expression.Expression, 0, len(conds))
+	selectedIncome := 0.0
+	selectedColumnCount := 0
+	selectedSelectivity := 1.0
+	totalColumnCount := len(physicalTableScan.Columns)
+
+	for _, exprGroup := range sortedConds {
+		mergedConds := append(selectedConds, exprGroup.exprs...)
+		selectivity, _, err := physicalTableScan.tblColHists.Selectivity(sctx, mergedConds, nil)
+		if err != nil {
+			logutil.BgLogger().Warn("calculate selectivity failed, do not push down the conditions group", zap.Error(err))
+			continue
+		}
+		var cols []*expression.Column
+		cols = expression.ExtractColumnsFromExpressions(cols, mergedConds, nil)
+		// Remove the duplicated columns
+		colSet := make(map[int64]struct{}, len(cols))
+		for _, col := range cols {
+			colSet[col.UniqueID] = struct{}{}
+		}
+		colCnt := len(colSet)
+		income := (1 - selectivity) * (float64(totalColumnCount) - float64(colCnt))
+		// If selectedColumnCount does not change,
+		// or the income increases after pushing down the group, push down it.
+		if colCnt == selectedColumnCount || income > selectedIncome {
+			selectedConds = mergedConds
+			selectedColumnCount = colCnt
+			selectedIncome = income
+			selectedSelectivity = selectivity
+		}
+		// If the selectivity improvement is small enough, break the loop
+		// to reduce the cost of calculating selectivity.
+		if selectedSelectivity-selectivity < selectivityImproveThreshold {
+			break
+		}
+	}
+
+	if len(selectedConds) == 0 {
+		return
+	}
+	logutil.BgLogger().Info("planner: push down conditions to table scan", zap.String("table", physicalTableScan.Table.Name.L), zap.String("conditions", expression.ExplainExpressionList(selectedConds, physicalSelection.Schema())))
```

Comment:
The info level is too high for this log, i think.

Reply:
change to debug